server/schedule/operator/step.go

```diff
@@ -325,7 +320,92 @@ func (bn BecomeNonWitness) Timeout(regionSize int64) time.Duration {
 
 // GetCmd returns the schedule command for heartbeat response.
 func (bn BecomeNonWitness) GetCmd(region *core.RegionInfo, useConfChangeV2 bool) *pdpb.RegionHeartbeatResponse {
-	return createResponse(addLearnerNode(bn.PeerID, bn.StoreID, false), useConfChangeV2)
+	return switchWitness(bn.PeerID, false)
+}
+
+// BatchSwitchWitness is an OpStep that batch switch witness.
+type BatchSwitchWitness struct {
+	ToWitnesses    []BecomeWitness
+	ToNonWitnesses []BecomeNonWitness
+}
+
+func (bsw BatchSwitchWitness) String() string {
+	b := &strings.Builder{}
+	_, _ = b.WriteString("batch switch witness")
+	for _, w := range bsw.ToWitnesses {
+		_, _ = fmt.Fprintf(b, ", switch peer %v on store %v to witness", w.PeerID, w.StoreID)
+	}
+	for _, nw := range bsw.ToNonWitnesses {
+		_, _ = fmt.Fprintf(b, ", switch peer %v on store %v to non-witness", nw.PeerID, nw.StoreID)
+	}
+	return b.String()
+}
+
+// ConfVerChanged returns the delta value for version increased by this step.
+func (bsw BatchSwitchWitness) ConfVerChanged(region *core.RegionInfo) uint64 {
+	return 0 // switch witness never change the conf version
```

Comment:
I think we should, otherwise, region cache won't be updated when witness is changed

Reply:
Yes, we should. I met this issue when testing TiKV with test_pd_client.